libsqlite3-sys/build.rs

```diff
@@ -593,47 +778,572 @@ mod bindings {
+                Some(version) => {
+                    let require_sqlite3_version = version_compare::Version::from(version)
+                        .expect("failed to parse required sqlite3 version as Version");
+                    let require_sqlite3_version_parts = require_sqlite3_version.parts();
+                    if require_sqlite3_version_parts.len() != 3 {
+                        panic!(
+                            "sqlite3 version '{}' does not have exactly three parts",
+                            version
+                        );
+                    }
+                    let major = match require_sqlite3_version_parts[0] {
+                        version_compare::Part::Number(major) => major,
+                        _ => {
+                            panic!(
+                                "non-numeric major part found in sqlite3 version requirement '{}'",
+                                require_sqlite3_version_parts[0]
+                            );
+                        }
+                    };
+                    let minor = match require_sqlite3_version_parts[1] {
+                        version_compare::Part::Number(minor) => minor,
+                        _ => {
+                            panic!(
+                                "non-numeric minor part found in sqlite3 version requirement '{}'",
+                                require_sqlite3_version_parts[1]
+                            );
+                        }
+                    };
+                    let patch = match require_sqlite3_version_parts[2] {
+                        version_compare::Part::Number(patch) => patch,
+                        _ => {
+                            panic!(
+                                "non-numeric patch part found in sqlite3 version requirement '{}'",
+                                require_sqlite3_version_parts[2]
+                            );
+                        }
+                    };
+                    let require_sqlite3_version_number = 1_000_000 * major + 1_000 * minor + patch;
+                    quote! {
+                        let sqlite3_version_number = sqlite3_libversion_number();
+                        if sqlite3_version_number < #require_sqlite3_version_number {
+                            panic!(stringify!("sqlite3 version is {} but version {} is required for the ", #api_fn_ident, " function"), sqlite3_version_number, #require_sqlite3_version_number);
+                        }
+                    }
+                }
+                None => quote! {},
+            };
+
+            // generate wrapper and return it as a string
+            let wrapper_tokens = quote! {
+                pub unsafe fn #api_fn_ident(#api_fn_inputs) #api_fn_output {
+                    let p_api = loadable_extension_sqlite3_api();
+                    #api_version_check_tokens
+                    ((*p_api).#field_ident
```

Comment:
Watch out: if `sqlite3_api_routines` is actually smaller than we think, then this becomes UB — it creates an intermediate reference, and references are marked as dereferencable (in llvm) for the whole size of the struct.

Instead you have to use `core::ptr::addr_of!((*p).field).read()`, for example.

Reply:
My intent was that the code inserted by #api_version_check_tokens applies the runtime version check and prevents the next line from being executed if `sqlite3_api_routines` is smaller at runtime than at compile time.

I'm not extremely well versed in this part of Rust. I guess the undefined behaviour is because we are never allowed to make a reference (not a raw pointer, but a reference) to uninitialized data even if we don't execute it, because the compiler may rearrange execution order or something like that? 

I'll make the change to raw pointers using `addr_of` and `read` - thanks for the suggestion!